src/Select.purs

```diff
@@ -90,266 +78,317 @@ derive instance ordVisibility :: Ord Visibility
+      -- | ```
+      itemProps :: _ -> _ -> _ -> Int -> Array (HP.IProp (ItemProps itemProps) (HookM m Unit))
+      itemProps getState modifyState pushSelectedIdxChanged index =
+        [ HE.onMouseDown \ev -> Just (select getState modifyState pushSelectedIdxChanged (Index index) (Just ev))
+        , HE.onMouseOver \_ -> Just (highlight getState modifyState (Index index))
+        ]
+
+      -- | An array of `IProps` with a `MouseDown`
+      -- | handler. It prevents clicking on an item within an enclosing HTML element
+      -- | from bubbling up a blur event to the DOM. This should be used on the parent
+      -- | element that contains your items.
+      containerProps :: Array (HP.IProp (onMouseDown :: ME.MouseEvent | containerProps) (HookM m Unit))
+      containerProps =
+        [ HE.onMouseDown \ev -> Just (preventMouseEvent ev) ]
+
+
+      -- | An array of `IProps` with `InputProps`. It
+      -- | allows the input element to capture string values, register key events for
+      -- | navigation, record open and close events based on focus and blur, and to be
+      -- | focused with the tab key.
+      -- |
+      -- | ```purescript
+      -- | renderInput = input_ (setInputProps [ class "my-class" ])
+      -- | ```
+      inputProps :: _ -> _ -> _ -> _ -> _ -> Array (HP.IProp (InputProps inputProps) (HookM m Unit))
+      inputProps getState modifyState searchDebouncer pushVisibilityChanged pushSelectedIdxChanged =
+        [ HE.onFocus \_ -> Just (setVisibility getState modifyState pushVisibilityChanged On)
+        , HE.onKeyDown \ev -> Just (key getState modifyState pushVisibilityChanged pushSelectedIdxChanged ev)
+        , HE.onValueInput \str -> Just (handleSearch getState modifyState searchDebouncer pushVisibilityChanged str)
+        , HE.onMouseDown \_ -> Just (setVisibility getState modifyState pushVisibilityChanged On)
+        , HE.onBlur \_ -> Just (setVisibility getState modifyState pushVisibilityChanged Off)
+        , HP.tabIndex 0
+        , HP.ref (H.RefLabel "select-input")
+        ]
+
+      getTargetIndex :: Maybe Int -> Int -> Target -> Int
+      getTargetIndex highlightedIndex itemCount = case _ of
+        Index i -> i
+        Prev -> case highlightedIndex of
+          Just i | i /= 0 -> i - 1
+          _ -> itemCount - 1
+        Next -> case highlightedIndex of
+          Just i | i /= (itemCount - 1) -> i + 1
+          _ -> 0
+
+      setFocus :: Boolean -> HookM m Unit
+      setFocus shouldFocus = do
+        inputElement <- Hooks.getHTMLElementRef $ H.RefLabel "select-input"
+        for_ inputElement \el -> H.liftEffect case shouldFocus of
+          true -> HTMLElement.focus el
+          _ -> HTMLElement.blur el
+
+      setVisibility
+        :: HookM m SelectState
+        -> ((SelectState -> SelectState) -> HookM m Unit)
+        -> (Visibility -> HookM m Unit)
+        -> Visibility
+        -> HookM m Unit
+      setVisibility getState modifyState pushVisibilityChanged v = do
```

Comment:
By moving these implementations into `let` you can skip most of these function arguments and avoid using `getState` altogether.

Reply:
I decided not to do that. The number of args passed to the function are smaller now because of the `inputRec` being in scope. I did try putting these back into the `let` statement, but it still seemed verbose.